File: web/src/api/ship.ts
import type { Target } from "../types/Target";
import type { Bundle } from "../types/Bundle";
import type { VerifyOutcome } from "../types/VerifyOutcome";
import { scopedPath } from "./client";

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) throw new Error(`${res.status}: ${await res.text()}`);
  return res.json() as Promise<T>;
}

export const listTargets = (pid: string) => fetch(scopedPath(pid, "/targets")).then(json<Target[]>);
export const listBundles = (pid: string) => fetch(scopedPath(pid, "/bundles")).then(json<Bundle[]>);
export const build = (pid: string, target: string) =>
  fetch(scopedPath(pid, "/build"), {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ target }),
  }).then(json<Bundle>);

export const verifyBundle = (pid: string, path: string) =>
  fetch(scopedPath(pid, "/verify"), {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ path }),
  }).then(json<VerifyOutcome>);

File: gateway/src/api/ship.rs
use axum::{http::StatusCode, Json};

use crate::api::scope::Scoped;
use crate::ship::{BuildRequest, Bundle, Target, VerifyOutcome, VerifyRequest};

pub async fn targets(Scoped(state): Scoped) -> Json<Vec<Target>> {
    Json(state.list_targets())
}

pub async fn bundles(Scoped(state): Scoped) -> Json<Vec<Bundle>> {
    Json(state.list_bundles())
}

pub async fn build(
    Scoped(state): Scoped,
    Json(req): Json<BuildRequest>,
) -> Result<Json<Bundle>, (StatusCode, String)> {
    state
        .build(&req.target)
        .map(Json)
        .map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))
}

pub async fn verify(
    Scoped(state): Scoped,
    Json(req): Json<VerifyRequest>,
) -> Result<Json<VerifyOutcome>, (StatusCode, String)> {
    state
        .verify(&req.path)
        .map(Json)
        .map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))
}
